terraform/data/bq_views/events/pull_request_events.sql

```diff
@@ -0,0 +1,64 @@
+  INT64(JSON_QUERY(PARSE_JSON(payload), "$.organization.id")) organization_id,
+  JSON_VALUE(PARSE_JSON(payload), "$.repository.full_name") repository_full_name,
+  INT64(JSON_QUERY(PARSE_JSON(payload), "$.repository.id")) repository_id,
+  JSON_VALUE(PARSE_JSON(payload), "$.repository.name") repository,
+  JSON_VALUE(PARSE_JSON(payload), "$.repository.visibility") repository_visibility,
+  JSON_VALUE(PARSE_JSON(payload), "$.sender.login") sender,
+  INT64(JSON_QUERY(PARSE_JSON(payload), "$.sender.id")) sender_id,
+  JSON_VALUE(PARSE_JSON(payload), "$.pull_request.active_lock_reason") active_lock_reason,
+  INT64(JSON_QUERY(PARSE_JSON(payload), "$.pull_request.additions")) additions,
+  JSON_VALUE(PARSE_JSON(payload), "$.pull_request.base.ref") base_ref,
+  INT64(JSON_QUERY(PARSE_JSON(payload), "$.pull_request.changed_files")) changed_files,
+  TIMESTAMP(JSON_VALUE(PARSE_JSON(payload), "$.pull_request.closed_at")) closed_at,
+  INT64(JSON_QUERY(PARSE_JSON(payload), "$.pull_request.comments")) comments,
+  INT64(JSON_QUERY(PARSE_JSON(payload), "$.pull_request.commits")) commits,
+  TIMESTAMP(JSON_VALUE(PARSE_JSON(payload), "$.pull_request.created_at")) created_at,
+  INT64(JSON_QUERY(PARSE_JSON(payload), "$.pull_request.deletions")) deletions,
+  BOOL(JSON_QUERY(PARSE_JSON(payload), "$.pull_request.draft")) draft,
+  JSON_VALUE(PARSE_JSON(payload), "$.pull_request.head.ref") head_ref,
+  JSON_VALUE(PARSE_JSON(payload), "$.pull_request.html_url") html_url,
+  INT64(JSON_QUERY(PARSE_JSON(payload), "$.pull_request.id")) id,
+  BOOL(JSON_QUERY(PARSE_JSON(payload), "$.pull_request.locked")) locked,
+  BOOL(JSON_QUERY(PARSE_JSON(payload), "$.pull_request.maintainer_can_modify")) maintainer_can_modify,
+  JSON_VALUE(PARSE_JSON(payload), "$.pull_request.merge_commit_sha") merge_commit_sha,
+  JSON_VALUE(PARSE_JSON(payload), "$.pull_request.mergeable_state") mergeable_state,
+  BOOL(JSON_QUERY(PARSE_JSON(payload), "$.pull_request.merged")) merged,
+  TIMESTAMP(JSON_VALUE(PARSE_JSON(payload), "$.pull_request.merged_at")) merged_at,
+  JSON_VALUE(PARSE_JSON(payload), "$.pull_request.merged_by.login") merged_by,
+  INT64(JSON_QUERY(PARSE_JSON(payload), "$.pull_request.number")) number,
+  INT64(JSON_QUERY(PARSE_JSON(payload), "$.pull_request.review_comments")) review_comments,
+  JSON_VALUE(PARSE_JSON(payload), "$.pull_request.state") state,
+  JSON_VALUE(PARSE_JSON(payload), "$.pull_request.title") title,
+  TIMESTAMP(JSON_VALUE(PARSE_JSON(payload), "$.pull_request.updated_at")) updated_at,
+  INT64(JSON_QUERY(PARSE_JSON(payload), "$.pull_request.user.id")) author_id,
+  JSON_VALUE(PARSE_JSON(payload), "$.pull_request.user.login") author,
+  TIMESTAMP_DIFF( TIMESTAMP(JSON_VALUE(PARSE_JSON(payload), "$.pull_request.closed_at")), TIMESTAMP(JSON_VALUE(PARSE_JSON(payload), "$.pull_request.created_at")), SECOND) open_duration_seconds
+FROM
+  `${dataset_id}.${table_id}`
+WHERE
+  event = "pull_request";
```

Comment:
I'll defer to @bradegler, but I think this is quite inefficient because we end up parsing the JSON 50 times. It might be better to parse the JSON as a CTE and then do a sub-SELECT.

Reply:
You dont need the parse, you can just do `JSON_VALUE(payload, "$.action") action`

Reply:
I was hoping query engine was smart enough to only do `PARSE_JSON(payload)` once even though it is stated several times. I think mike is write that it is not strictly needed though.

Reply:
I just want to make sure `JSON_VALUE` isn't going to parse the substantially large payload N times for each row.

Reply:
`JSON_VALUE` is not sufficient as it returns a string https://cloud.google.com/bigquery/docs/reference/standard-sql/json_functions#json_value. 

We don't want every column to be a string, but parsing funcitions such as `INT64` take a a `json_expr` and not a string. So we do actutally need to parse the JSON.

Reply:
I updated to parse the json in a subquery.

Reply:
I dont think we need to do this, JSON_VALUE is efficient at getting the values out. If you need to cast to another type you can do the following for the type you want, specifically INT64 and BOOL:

```
  SAFE_CAST(JSON_VALUE(payload, "$.organization.id") AS INT64) organization_id
```

Reply:
How about I just have the `unique_events.sql` view parse the json. Then every view derived from it doesn't need to worry about parsing the json. I think it is valuable having the json parsed as then the full set of JSON fucntions big query provides can be used when creating views/metrics.

Reply:
actually my suggestion doesn't work as the unique events view groups by the payload, and JSON values are not allowed in group by expressions apparently.

Reply:
I tried re-writing this with a CTE and its actually less performant, both speed and data size. I would keep this as is, update to use SAFE_CAST as needed for data types as mentioned.

Reply:
Mike's suggestion was more performant, So I've updated to use `SAFE_CAST` where needed as per his suggestion.